[1단계 - 출석 구현] 멍구(이유영) 미션 제출합니다.

yuyoungrhee opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->

## 체크 리스트

- [x] 미션의 필수 요구사항을 모두 구현했나요?
    - 프로그래밍 요구사항 일부는 지키지 못했습니다: 메서드 길이 10줄, depth 1 제한 등
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
    - https://prolog.techcourse.co.kr/studylogs/3956


## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?

### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [x] 2
- [ ] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
- [ ] 한 메서드에 오직 한 단계의 들여쓰기(indent)만 한다.
- [x] else 예약어를 쓰지 않는다.
- [x] 모든 원시값과 문자열을 포장한다.
    - 대부분 포장했으나 LocalDateTime 등 "출석시간"을 나타내는 원시값을 포장하지 않음
- [ ] 한 줄에 점을 하나만 찍는다.
- [x] 줄여쓰지 않는다(축약 금지).
- [ ] 모든 엔티티를 작게 유지한다.
    - AttendanceHistory가 조금 큰 것 같다.
- [x] 3개 이상의 인스턴스 변수를 가진 클래스를 쓰지 않는다.
- [x] 일급 콜렉션을 쓴다.
- [ ]: 게터/세터/프로퍼티를 쓰지 않는다.
    - 크루 안에 있는 기록 안에 잇는 기록상세 이렇게 구현하다보니 getter를 사용해서 우선 구현. getter를 없앨 수 있도록 리팩터링을 완료하지 못함

  
## 어떤 부분에 집중하여 리뷰해야 할까요?

<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 피드백받고 싶은 내용을 공유해주세요.  
예를 들어, 가장 고민했던 점이나 여전히 어려운 부분, 그리고 이에 대한 생각을 적을 수 있습니다. -->
### 책임 설계
- Crews -> Crew -> History -> Detail로 연쇄적으로 의존이 이어지는 구조인데, 다 각자 해당하는 책임을 수행하려면 서로의 필드를 가지고있을 수밖에 없었습니다. 이 상태에서 어떻게 책임을 잘 설계할 수 있는지 궁금합니다. 

### TDD
- TDD를 할 때 해당 기능에 대해서 먼저 테스트를 구현하다 보니 기능이 어떤식으로 동작하는지 예측하기가 조금 어려웠습니다. 그래서 실제로 테스트 작성 -> 기능작성 -> 빼먹은 테스트 작성 이런식으로 진행될 때가 있었는데 기능에 대해 더 설계를 생각해보고 구현을 진행하는 게 좋을까요?

### CustomLocalDateTime
요구사항이 2024년 12월 한달 기준이어서 실제 날짜로 테스트할 수 없는 상황이라 우선 2024년 12월 16일 날짜로 실행할 수 있도록 코드를 구성했습니다. 더 좋은 방법 있을까요? 해당 방법이 괜찮은지 고민이었습니다

yuyoungrhee committed:
docs: 기능 요구사항 작성

yuyoungrhee committed:
feat: 크루 테스트 작성

yuyoungrhee committed:
feat: 출결 기능 구현

yuyoungrhee committed:
test: 요일 시작 시간 확인 테스트 작성

yuyoungrhee committed:
feat: 요일 시작 시간 model 작성

yuyoungrhee committed:
test: 요일 종료시간 테스트 작성

yuyoungrhee committed:
test: 요일 종료시간 model로직 작성

yuyoungrhee committed:
feat: Attendence enum에서 출석한 시간을 통해서 객체를 생성하도록 수정

- 기존 int time에서 LocalDateTime으로 인자 변경, 출석한 시간으로 생성할 수 있도록 함

yuyoungrhee committed:
feat: 경고 대상자를 나타내는 기능 작성

yuyoungrhee committed:
feat: 출석 기록에서 유지하는 출석 상세 구현

yuyoungrhee committed:
feat: 출석 기록을 유지하는 기능 구현

yuyoungrhee committed:
feat: AttendenceDetail 수정 기능 추가

yuyoungrhee committed:
feat: Crew가 출석 기록을 가지도록 수정

yuyoungrhee committed:
chore: Attendence -> Attendance로 오탈자 수정

yuyoungrhee committed:
feat: 크루들을 저장하는 Crews 작성

yuyoungrhee committed:
feat: Attendances.csv파일을 읽어서 Crews에 저장하는 로직 작성

yuyoungrhee committed:
feat: view 작성

yuyoungrhee committed:
fix: 같은 Crew가 여러명이 들어가는 오류 수정

yuyoungrhee committed:
test: CrewDataLoader 변경에 따른 테스트 수정

yuyoungrhee committed:
feat: Controller + 메인 구현

yuyoungrhee committed:
feat: 오늘 날짜를 원하는 날짜로 주입받을 수 있도록 기능 구현

yuyoungrhee committed:
feat: 오늘이 공휴일인지 확인하는 메서드 작성

yuyoungrhee committed:
feat: 등교하지 않은 날은 결석으로 처리

yuyoungrhee committed:
fix(AttendanceDetail): 출석 수정이 가능하므로 final 삭제

yuyoungrhee committed:
feat: 출석 수정 기능 구현

yuyoungrhee committed:
feat: 제적 위험자 조회 기능 추가

yuyoungrhee committed:
fix: 출결이 없는 날짜는 결석 처리 되도록 수정

yuyoungrhee committed:
refactor: dto 적용, 출력 정렬

yuyoungrhee committed:
feat: 등록되지않은 크루를 조회시 예외처리 작성

yuyoungrhee committed:
feat: 공휴일에 대한 출결 수정시 예외처리

yuyoungrhee committed:
refactor: enum 멤버변수를 사용하도록 로직 수정

yuyoungrhee committed:
refactor: 예외처리 위치 변경

yuyoungrhee committed:
refactor: controller파싱 모듈화

yuyoungrhee committed:
refactor: controller 모듈화

yuyoungrhee committed:
refactor: 결석 계산 로직을 AttendanceWarning으로 위치 변경

yuyoungrhee committed:
refactor: 미사용 로직 삭제

yuyoungrhee committed:
refactor: 메서드 인자 변경 및 메서드명 변경 등

yuyoungrhee committed:
refactor: CrewDataLoader 메서드 분리

yuyoungrhee committed:
refactor: 출석 추가 메서드명 append로 변경

Commit:
test: 테스트 이름 작성

Commit:
test: 테스트 이름 작성

yuyoungrhee committed:
test: 테스트 추가 및 테스트 이름 작성

Commit:
refactor: 매직넘버 상수화

yuyoungrhee committed:
fix: 등교일이 아닐 때 출석 확인 기능 선택 시 예외가 바로 터지도록 수정

yuyoungrhee committed:
test: 테스트 이름 수정 및 테스트 추가

yuyoungrhee committed:
refactor: get으로 출석을 가져오지 않고 출석 상태를 비교하는 메서드를 사용하도록 수정

yebink commented on src/main/java/attendance/controller/Controller.java:
```diff
@@ -0,0 +1,102 @@
+        crewDataLoader.load(FILE_NAME);
+    }
+
+    public void run() {
+        String s = inputView.inputCommand();
+        if (s.equals("1")) {
+            processAddAttendance();
+        }
+        if (s.equals("2")) {
+            processModifyAttendance();
+        }
+        if (s.equals("3")) {
+            processDisplayAttendanceHistory();
+        }
+        if (s.equals(("4"))) {
+            processDisplayWarningCrew();
+        }
+        if (s.equals("Q")) {
+            System.exit(1);
+        }
+        run();
+    }
```
`함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.` 요구사항을 지켜볼까요~?

해당 내용 전반적으로 체크해주세요!

yuyoungrhee replied:
확인했습니다! 전반적으로 체크했는데, 놓친 부분이 있다면 다시 피드백 부탁드립니다..!!

yebink commented on src/main/java/attendance/controller/Controller.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public void run() {
+        String s = inputView.inputCommand();
```
`s` 보다 더 적절한 네이밍을 고려해봐요!

yuyoungrhee replied:
해당 변수를 사용하는 로직을 삭제해서 변경하진 못했지만, 만약에 그대로 로직을 유지한다면 네이밍은 command 정도로 했을 것 같습니다. 급했는지 s로 남겨두었네요..!

yebink commented on src/main/java/attendance/model/CustomLocalDateTime.java:
```diff
@@ -0,0 +1,44 @@
+
+public class CustomLocalDateTime {
+    public static final LocalDateTime FIXED_DATE_TIME = LocalDateTime.of(2024, 12, 16, 12, 0);
+    private static final Clock clock = Clock.fixed(
```
UTC 를 강제하신 이유가 있을까요?

yuyoungrhee replied:
원하는 시간을 지정하는 방법에 대해서 서칭 후 사용한 것인데, 아무래도 급하게 사용하다보니 충분한 이해 없이 사용하게된 것 같습니다.. UTC는 국제 표준 시간으로 설정하는 것이고, 한국에서만 사용하는 경우 `Asia/Seoul` 과 같이 ZoneId를 설정하면 된다고 합니다.! CustomClock 코드에서 반영했습니다.

yebink commented on src/main/java/attendance/model/CustomLocalDateTime.java:
```diff
@@ -0,0 +1,44 @@
+import java.time.format.DateTimeFormatter;
+import java.util.Set;
+
+public class CustomLocalDateTime {
```
객체로 관리할만한 상태와 util성 메서드가 섞여있는 것 같네요!
둘을 분리해볼까요~?

yuyoungrhee replied:
CustomClock 과 DateUtil로 분리해봤습니다!

yebink commented on src/main/java/attendance/dto/AttendanceDTO.java:
```diff
@@ -0,0 +1,40 @@
+import java.time.LocalDateTime;
+import java.util.List;
+
+public record AttendanceDTO(
```
보통 Dto 네이밍은 아래와 같이 합니다 😄 

```suggestion
public record AttendanceDto(
```

HttpUrlCsvDto 와 같은 축약 형태로만 이루어진 네이밍의 경우 어떻게 될까요?

yuyoungrhee replied:
모두 대문자로 표기하는 경우도 있어서 크게 신경쓰지 않았는데 3글자 이상의 축약어의 경우에는 앞글자만 대문자로 표기한다는 컨벤션이 일반적임을 알았습니다. VO같은 경우 2글자라서 모두 대문자로 써도 된다는 것도 알게 되었어요..!

컨벤션은 일관성이 중요한데, HttpUrlCsvDto는 모두 축약어이기 때문에 DTO를 모두 대문자로 표기한다고 정하게 된다면 HTTPURLCSVDTO라는 가독성 없는 클래스명이 탄생할 수도 있겠네요..😅

yebink commented on src/main/java/attendance/model/Attendance.java:
```diff
@@ -0,0 +1,27 @@
+import java.time.LocalDateTime;
+
+public enum Attendance {
+    출석(0),
```
일반적으로 변수명은 영문을 사용합니다!

한글은 자동완성이 안 되고, 일부 환경에서 제대로 동작하지 않을 여지가 있습니다. 

또, `출석` 이라는 것을 attendance 를 표현하는 하나의 용어로 본다면, 예를 들어, 멍구의 프로그램을 일본에서도 사용하기로 했다면 여전히 해당 변수명을 사용할 수 있을까요?

yuyoungrhee replied:
빠르게 개발하기 위해서 페어와 서로 이해하기 쉬운 한글로 변수명을 우선 작성하였습니다. 나중에 변경하고자 하였는데 리팩터링 과정에서 놓친 것 같습니다. 수정했습니다!

yebink commented on src/main/java/attendance/model/Crew.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    private void validate(String name) {
+        if (name == null || name.isBlank()) {
+            throw new IllegalArgumentException("크루의 이름은 1자 이상 5자 이하여야 합니다.");
+        }
+        if (name.length() >= 5) {
+            throw new IllegalArgumentException("크루의 이름은 1자 이상 5자 이하여야 합니다.");
+        }
```
if 문 2개에서 동일한 역할을 하고 있는데요!

두 개를 나누신 이유가 있을까요?

yuyoungrhee replied:
닉네임 입력이 비어있거나 null인 경우에는 이름의 길이가 1자 이상 되지 않는다고 생각하고 "길이"에 대한 예외 메시지를 던지도록 했습니다. 

굳이 예외 메시지를 나누지 않은 이유는 "크루의 이름은 1자 이상이어야합니다" 와 "크루의 이름은 5자 이하여야합니다"의 예외를 구분할 필요가 없다고 생각했기 때문입니다. 사용자 입장에서 같은 예외 메시지를 보고도 1자이상 입력을 하지 않았기 때문에 예외가 발생했는지, 5자 초과하여 입력을 하였는지 판단할 수 있기 때문이에요.

대신에 같은 예외 메시지에 대한 조건이므로 3개의 조건을 합치도록 변경했습니다. 
\+ 추가로 5자 이하인데 조건은 4자이하인 점을 놓쳐서 조건과 예외메시지를 약간 변경했습니다.
```java
if (name == null || name.isBlank() || name.length() > 4) {
            throw new IllegalArgumentException("크루의 이름은 1자 이상 4자 이하여야 합니다.");
}
```

yebink commented on src/test/java/input/LoadCrewDataTest.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    @Test
+    void 파일의_출석정보를_불러온다() {
+        loader.load("attendances.csv");
+        assertThat(crews.findCrew("빙티").getAttendanceHistory().stream())
+                .hasSize(10);
+    }
```
`// given`, `// when`, `// then` 주석을 이용해 가독성을 높여볼까요~?
테스트 전반적으로 반영해주세요 🙂

yuyoungrhee replied:
반영했습니다! 원래 해당 주석을 사용하는 편인데 테스트 리팩터링 시간이 부족해 적용하지 못했습니다😔

yebink commented on src/main/java/attendance/model/AttendanceDetail.java:
```diff
@@ -0,0 +1,50 @@
+import java.time.format.DateTimeFormatter;
+
+public class AttendanceDetail {
+    private final DateTimeFormatter formatter = DateTimeFormatter.ofPattern("MM월 dd일 EEEE은 등교일이 아닙니다.");
```
에러메시지가 formatter 에 그대로 정의되어 있네요!
재사용하기 좋은 포맷으로 변경해볼까요?

yebink replied:
AttendanceDetail 이 생성되는 갯수만큼 formatter도 같이 생성되겠네요!

그래서 보통 이런 메서드는 util 클래스로 분류합니다 🙂

yebink replied:
위 두가지 내용 코드 전체적으로 검토해주세요 🙂

yuyoungrhee replied:
반영했습니다! formatter를 인스턴스 변수로 선언한다면 객체가 생성될 때마다 중복 생성될 수 있겠네요.. util 클래스로 분리하여 적용했습니다.

yebink commented on src/main/java/attendance/model/AttendanceDetail.java:
```diff
@@ -0,0 +1,50 @@
+        this.attendance = Attendance.from(localDateTime);
+    }
+
+    private void validateHoliday(LocalDateTime localDateTime) {
```
메서드명만 보면 holiday 인지를 검증하는 것 같은데, 내부 로직은 주말인지 여부도 보고있군요! 적절한 메서드명을 고민해볼까요? 🤔

yuyoungrhee replied:
`valdiateWeekendOrHoliday`로 변경했습니다! 출석을 받지 않는 날이라는 메서드명도 고려해보았지만 일단은 가장 직관적인 메서드명으로 변경해보았습니다.

yebink commented on src/test/java/model/CrewsTest.java:
```diff
@@ -0,0 +1,33 @@
+    @Test
+    void 등록되지_않은_닉네임으로_찾으면_예외가_발생한다() {
+        Crews crews = new Crews();
+        assertThatThrownBy(() -> crews.findCrew("빙티"))
+                .isInstanceOf(IllegalArgumentException.class);
```
exception 테스트를 할 때에는 메시지도 꼭 함께 테스트해주세요!

```suggestion
        assertThatThrownBy(() -> crews.findCrew("빙티"))
                .isInstanceOf(IllegalArgumentException.class)
                .hasMessageContaining("등록되지 않은 닉네임입니다.");
```

yuyoungrhee replied:
모두 수정했습니다! 메시지를 함께 테스트하지 않는다면 예상과는 다른 예외로 IllegalArgumentException이 터져도 모를 수 있기 때문에 검증해야 한다는 점에서 필요하군요

yebink commented on src/main/java/attendance/model/CrewDataLoader.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Objects;
+
+public class CrewDataLoader {
+    public static final LocalTime ABSENCE_TIME = LocalTime.of(17, 0);
```
임의의 숫자를 지정한 이유는 무엇일까요?
더 좋은 방법은 없을까요?

yebink commented on src/main/java/attendance/model/CrewDataLoader.java:
```diff
@@ -0,0 +1,82 @@
+        }
+    }
+
+    private boolean isAttendableDate(LocalDate currentDate) {
```
파라미터로 들어온 currentDate 가 1월 1일보다 작은지, 12월 16일보다 작은지 를 둘 다 체크하고 있네요! 불필요한 부분을 제거해볼까요?

yebink commented on src/main/java/attendance/model/CrewDataLoader.java:
```diff
@@ -0,0 +1,82 @@
+        fillAbsencesForNoAttendance();
+    }
+
+    private void fillAbsencesForNoAttendance() {
```
크루마다 출석 정보가 없는 날짜에 대해 결석을 적재하고 있네요.

크루가 100만 명이 되면 어떨까요?
날짜가 1달이 아니라 10년이 되면 어떨까요?
매번 결석 을 저장하는 건 낭비가 아닐까요~?

좀 더 확장성을 고려해서 구현해봅시다!

yuyoungrhee replied:
CrewDataLoader에 대한 코멘트에 대해서 한 번에 답변 드리겠습니다!

최대한 쉬운 방향으로 우선 기능을 완성하고자 결석날짜와 시간을 임의로 지정하여 결석 기록 객체를 새로 생성하는 방식으로 구현했었습니다. 코멘트와 같이 자원 낭비 방지와 확장성을 고려햐여 다음과 같이 수정했습니다!

**수정 후**
- `AttendanceReport` 객체를 새로 구현하였고, 해당 객체에서 **교육 시작일**과 **시계**(오늘 날짜를 가르킴)를 주입받아 `기록 결석일` + `출석기록이 없는 결석일`을 계산했습니다.
- 출력 시 결석일을 따로 표기해주는 것은 outputView에서 따로 없는 날짜에 대한 적절한 처리 후 출력하도록 했습니다.

yebink commented on src/main/java/attendance/model/AttendanceDetail.java:
```diff
@@ -0,0 +1,50 @@
+    }
+
+    public boolean isSameAs(Attendance attendance) {
+        return attendance.equals(this.attendance);
```
보통 Enum 비교는 `equals` 보단 `==` 을 씁니다~

```suggestion
        return attendance == this.attendance;
```

yuyoungrhee replied:
수정했습니다! 몰랐던 내용이네요. 

- Enum은 내부적으로 싱글톤 관리이기 때문에 ==를 사용해도 안전하게 비교할 수 있다
- 성능도 equals보다 좋다 (equals는 내부적으로 instanceof 확인 후 비교, ==는 즉시 참조 비교)
- NPE가 발생하지 않는다
- == 비교는 컴파일 타임에 타입 호환성 검사를 지원하기 때문에 프로그래머의 실수를 컴파일 수준에서 빠르게 잡을 수 있다

yebink commented on src/main/java/attendance/view/CustomStringBuilder.java:
```diff
@@ -0,0 +1,13 @@
+package attendance.view;
+
+public class CustomStringBuilder {
```
👍

yebink commented on src/main/java/attendance/view/CustomStringBuilder.java:
```diff
@@ -0,0 +1,13 @@
+    private final StringBuilder stringBuilder = new StringBuilder();
+
+    public void appendLine(String content) {
+        stringBuilder.append(content).append('\n');
```
```suggestion
        stringBuilder.append(content).append(System.lineSeparator());
```

`System.lineSeparator()` 를 사용했을 때 어떤 이점이 있을까요?

yuyoungrhee replied:
시스템에 따라서 사용하는 개행문자가 다르기 때문에 사용하는군요!

- 문자열 사이에 함부로 escape 기호인 \n을 사용하는 것은 다른 종류의 OS에서 예상하지 못한 동작이 될 수 있기 때문입니다.
- 자바에서 대부분의 입출력API의 경우 \n을 자동으로 OS에 맞게 변환해주지만, Windows환경에서 일부 프로그램의 경우 줄바꿈이 적용이 되지 않을 수 있습니다. (예: 코드 에디터가 아닌 다른 프로그램 등)

yebink commented on src/main/java/attendance/model/CrewDataLoader.java:
```diff
@@ -0,0 +1,82 @@
+import java.time.format.DateTimeFormatter;
+import java.util.Objects;
+
+public class CrewDataLoader {
```
해당 클래스의 책임을 좀 줄여볼까요?
객체지향 5대원칙 중 SRP 원칙을 지키며 구현해봅시다!

yuyoungrhee replied:
파일에서 데이터를 읽기 + 객체 생성 + 객체에 정보 추가 라는 책임을 loader 패키지 안의 클래스들로 적절히 분리해봤습니다! 

AttendanceLoader 에서 순수 파일의 정보를 읽어오고, AttendanceProcessor에서 파일 정보 -> 객체 변경 및 정보추가 처리 하도록 분리해보았습니다..!

yebink reviewed (changes requested):
안녕하세요 멍구~

늦어서 죄송합니다 !
출석 미션 구현하느라 고생 많으셨습니다 😄 💯 

피드백 몇가지 드렸으니 확인 후 다시 요청 주세요~

yebink commented:
안녕하세요 멍구~

주신 질문에 대한 답변 드립니다.

**1. 책임 설계**
* 객체의 의존방향까지 고민하셨군요 👍 
* 객체 간 상호 참조가 발생하는 것이 아니기 때문에 지금 코드도 좋다고 생각합니다.
* 연쇄적인 의존이 발생한 사유를 말씀드리자면, 멍구의 코드에는 크게 `크루` / `출석` 도메인이 있고, `크루` 도메인에 `출석` 도메인이 완전히 의존하고 있는 상태입니다. (`출석` 도메인이 `크루` 도메인 객체의 생애주기를 따라갑니다.)
* 연쇄 의존을 해소해보고 싶다면, `Crew` 에서 `AttendanceHistory` 를 제거해보세요.

**2. TDD**
* TDD 사이클을 경험하다보면 흔히 마주치는 일입니다 👏 
* 멍구의 커밋을 보니 요구사항을 리스트업하고, 가장 작은 단위부터 테스트를 작성하셨네요.
* step2 에서는 요구사항을 보고 어느정도 객체 간 상호작용 등을 고민해본 다음 테스트를 짜볼까요?
* 정답은 없는 문제이니 다양한 방법을 경험해보시고, 가장 편리한 방법을 찾으시면 좋겠습니다 😄 

**3. CustomLocalDateTime**
* 네 지금 방식도 괜찮습니다!
* 더 욕심이 나신다면, 사용자에게 날짜를 입력받는 방식은 어떨까요~?
* 1~31 일 날짜를 입력받고, 그 외의 날짜는 예외를 터뜨리는 방식도 있겠네요.
* 필수 요구사항은 아니니 굳이 구현하지 않으셔도 됩니다 ㅎㅎ

고생 많으셨습니다.
전반적으로 단위테스트가 매우 꼼꼼하네요 💯 👍

yuyoungrhee committed:
refactor(Controller): 메서드 10줄 이하로 수정

- 커맨드와 실행할 메서드를 Map으로 관리하여 메서드를 10줄 이하로 줄입

yuyoungrhee committed:
refactor: enum 변수명 영어로 변경 및 메서드 10줄 이하 지키도록 수정

yuyoungrhee committed:
refactor(Crew): 검증 로직 합침 및 검증 로직 조건 수정

yuyoungrhee committed:
refactor: 축약어의 앞글자만 대문자로 표시하도록 수정 DTO -> Dto

yuyoungrhee committed:
refactor(error): 예외 메시지 상수화 및 테스트 시 예외 메시지 검증

yuyoungrhee committed:
refactor: 테스트 패키지 이동

yuyoungrhee committed:
refactor: 테스트 시나리오를 given-when-then 형식으로 작성

yuyoungrhee committed:
refactor: 날짜 포맷 기능을 util 클래스로 분리하여 적용

yuyoungrhee committed:
refactor: 출석을 받지않는 주말+공휴일을 검증하는 메서드명 수정

yuyoungrhee committed:
refactor: enum 비교 equals -> ==

yuyoungrhee committed:
refactor: 데이터 로드 클래스 책임 분리 및 LocalDate 원시값 포장

yuyoungrhee committed:
refactor(loader): 데이터 로더를 책임지는 클래스들 역할 재분배

- 사용하지 않는 클래스 삭제

yuyoungrhee committed:
refactor(WoowaDate): 필요없는 코드 삭제

yuyoungrhee committed:
refactor(CustomStringBuilder): System.lineSeperator 사용

yuyoungrhee committed:
refactor(Controller): 메서드 분리

yuyoungrhee committed:
refactor: 결석 기록을 추가하지 않고 계산하는 방법으로 수정

AttendanceHistory에 출석 기록이 없을 때 결석한 기록을 생성하여 추가하는 방법이 아니라 결석일수를 계산하는 방법으로 변경함
- AttendanceReport를 통해 교육시작일로부터 오늘날짜까지 출석기록이 없는 경우 결석으로 처리하도록 하여 결석 횟수를 계산
- OutputView에서 직접 출석기록이 없는 경우 --:--로 출력하도록 함

yuyoungrhee committed:
refactor: 사용하지 않는 클래스 삭제 및 메서드 순서 변경

yuyoungrhee committed:
refactor(FixedCustomClock): 날짜를 외부에서 설정할 수 있도록 변경

yuyoungrhee committed:
test(TestUtil): 테스트를 위한 CustomClock 설정 유틸 추가

yuyoungrhee committed:
test: 빠진 테스트 추가

yuyoungrhee committed:
refactor(Crews): Crews를 반환하는 getter가 불변리스트를 반환하도록 수정

yuyoungrhee committed:
refactor(AttendanceProcessor): CustomClock 인터페이스를 사용하도록 수정

yuyoungrhee committed:
refactor(AttendanceLoader): 내부에서만 사용하고 있는 메서드 private으로 변경

yuyoungrhee committed:
refactor: enum 영어식별자로 변경, enum 한글 명이 출력되도록 수정

yuyoungrhee commented:
안녕하세요 엘리😀 남겨주신 코멘트 도움 많이 됐습니다. 아래는 큰 구조 상 수정한 내용들입니다.

### 1. CrewDataLoader 책임 분리
파일 데이터 읽기 / Crews 객체 관리 및 초기화 / 각 Crew의 기록에 파일데이터 저장

### 2. LocalDate 값 포장
WoowaDate라는 LocalDate를 감싸는 객체 구현, 오늘이 교육일인지 검증 후 객체가 생성되도록 함

### 3. Util 클래스 분리
DateUtil  - 기본적인 parse와 해당 요일이 주말인지 판단
CustomClock - 오늘 날짜 관리 및 제공, 공휴일 관리

### 4. 결석 기록 추가 방식 변경
- 기존: 결석 기록을 크루 기록에 직접 AttendanceDetail 객체를 추가하는 방식
- 변경: 교육 시작 날짜와 오늘 날짜에 기반하여 결석 횟수 계산, 출력 시 기록이 없을 경우 --:-- 형식으로 출력하도록 추가 구현

yebink commented on src/test/java/attendance/model/AttendanceReportTest.java:
```diff
@@ -0,0 +1,70 @@
+
+        AttendanceReport report = new AttendanceReport(TestUtil.getClock(), attendanceHistory,
+                TestUtil.getTrainingStartDate());
+        //then
```
이렇게 when 절이 별도로 선언되지 않는 경우, 
보통 when 주석을 생략하기보단`// when & then` 을 함께 씁니다 🙂

yebink commented on src/main/java/attendance/model/WoowaDate.java:
```diff
@@ -0,0 +1,33 @@
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+
+public class WoowaDate {
```
👍

yebink commented on src/main/java/attendance/util/DateUtil.java:
```diff
@@ -0,0 +1,25 @@
+import java.time.DayOfWeek;
+import java.time.LocalDate;
+
+public class DateUtil {
```
util 클래스에는 보통 어떠한 비즈니스 로직과도 연관이 없는 클래스들을 모아둡니다.
`parseDate` 메서드에서는 날짜 day를 입력받아 2024년 12월 day일 을 리턴하고 있는데요~

2024년 12월 이라는 것도 결국 비즈니스 요구사항과 밀접한 관련이 있겠죠~?

`java.util` 패키지를 참고해 util성 메서드와 아닌 메서드를 구분해 보시고, step2 에서 둘을 잘 분리해보아요 😄

yebink commented on src/main/java/attendance/controller/Controller.java:
```diff
@@ -0,0 +1,113 @@
+
+    public void run() {
+
+        Map<String, Runnable> commands = Map.of("1", this::processAddAttendance, "2", this::processModifyAttendance,
+                "3", this::processDisplayAttendanceHistory, "4", this::processDisplayWarningCrew, "Q",
+                () -> System.exit(0));
```
`1`, `2`, `3`, `4`, `Q` 와 같은 명령어를 Enum으로 선언해보면 어떨까요~?

yebink commented on src/main/java/attendance/util/DateFormatUtil.java:
```diff
@@ -0,0 +1,31 @@
+
+public class DateFormatUtil {
+
+    public static final DateTimeFormatter NOT_ATTENDABLE_FORMATTER = DateTimeFormatter.ofPattern(
```
해당 포매터에서 `MM월 dd일 EEEE` 형태로 포매팅을 한다면, InputView 에서도 재사용이 가능하겠죠~?

yebink commented on src/main/java/attendance/model/FixedCustomClock.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.ZoneId;
+import java.util.Set;
+
+public class FixedCustomClock implements CustomClock {
```
👍

yebink commented on src/main/java/attendance/model/loader/AttendanceLoader.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.Objects;
+
+public class AttendanceLoader {
+    public static final String FILE_NAME = "attendances.csv";
+    public static final String ROW_DELIMITER = ",";
+    public static final int CREW_NAME_COLUMN_INDEX = 0;
+    public static final int ATTENDANCE_DATE_TIME_COLUMN_INDEX = 1;
```
private 으로 선언해도 되겠네요 😄 💯

yebink commented on src/main/java/attendance/model/loader/AttendanceLoader.java:
```diff
@@ -0,0 +1,45 @@
+import java.util.List;
+import java.util.Objects;
+
+public class AttendanceLoader {
```
👍

yebink reviewed (approved):
안녕하세요 멍구~

피드백 빠르게 잘 반영해주셨네요 💯 
몇가지 추가 코멘트 드렸으니 확인해보시고 step2 에서 반영해주세요!

머지하겠습니다 😄
